pulp/apis/cuopt_api.py

```diff
@@ -0,0 +1,337 @@
+                    var.dj = value
+
+                duals = solution.get_dual_solution()
+                for constr, value in zip(lp.constraints.values(), duals):
+                    constr.pi = value
+
+            return status
+
+        def available(self):
+            """True if the solver is available"""
+            return True
+
+        def callSolver(self, lp, callback=None):
+            """Solves the problem with CUOPT"""
+            self.solveTime = -clock()
+            # TODO: Add callback
+            log_file = self.optionsDict.get("logPath") or ""
+
+            self.settings.set_infeasibility_detection(True)
+            self.solution = self.solver.Solve(lp.solverModel, self.settings, log_file)
+
+            self.solveTime += clock()
+
+        def buildSolverModel(self, lp):
+            """
+            Takes the pulp lp model and translates it into a COPT model
+            """
+            lp.solverModel = self.model
+
+            if lp.sense == LpMaximize:
+                lp.solverModel.set_maximize(True)
+            if self.timeLimit:
+                self.settings.set_time_limit(self.timeLimit)
+
+            gapRel = self.optionsDict.get("gapRel")
+            if gapRel:
+                self.settings.set_relative_gap_tolerance(gapRel)
+
+            var_lb, var_ub, var_type, var_name = [], [], [], []
+            obj_coeff = []
+            var_dict = {}
+
+            for i, var in enumerate(lp.variables()):
+                obj_coeff.append(lp.objective.get(var, 0.0))
+                lowBound = var.lowBound
+                if lowBound is None:
+                    lowBound = -np.inf
+                upBound = var.upBound
+                if upBound is None:
+                    upBound = np.inf
+                varType = "C"
+                if var.cat == LpInteger and self.mip:
+                    varType = "I"
+                if var.cat == LpBinary and self.mip:
+                    varType = "I"
+                    lowBound = 0
+                    upBound  = 1
+                var_lb.append(lowBound)
+                var_ub.append(upBound)
+                var_type.append(varType)
+                var_name.append(var.name)
+                var_dict[var.name] = i
+                var.solverVar = {var.name: {"lb": var_lb, "ub": var_ub, "type": var_type}}
+
+            lp.solverModel.set_variable_lower_bounds(np.array(var_lb))
+            lp.solverModel.set_variable_upper_bounds(np.array(var_ub))
+            lp.solverModel.set_variable_types(np.array(var_type))
+            lp.solverModel.set_variable_names(np.array(var_name))
+
+            var_warmstart = []
+            if self.optionsDict.get("warmStart", False):
+                for var in lp._variables:
+                    if var.varValue is not None:
+                        var_warmstart.append(var.varValue)
+                    else:
+                        var_warmstart = None
+                        break
```

Comment:
does this work? I'm not sure this does what it's supposed to do. If one variable has a None value then the list is incomplete.

Reply:
Removed for now. Might add back in future.